src/clara_specs/core.clj

```diff
@@ -8,46 +8,61 @@
+                       (s/cat :test #{:test}
+                              :expr (s/* any?))))
+
+(s/def ::constraint any?)
+
+(s/def ::condition (s/and vector?
+                          (s/cat :type any?
+                                 :constraints (s/* ::constraint))))
+
+
+(s/def ::accumulator-expr (s/and vector?
+                                 (s/cat :result-binding ::variable-name
+                                        :<- #{'<-}
+                                        :accumulator any?
+                                        :from #{:from}
+                                        :condition ::condition)))
 
 (s/def ::boolean-expr (s/and vector?
-                             (s/cat :condition #{:and :or :not}
-                                    :expression (s/+ (s/or :boolean-expr ::boolean-expr
-                                                           :fact-expr ::fact-expr))))) ; TODO what do these look like?
-
-(s/def ::fact-expr (s/and vector?
-                          (s/cat :binding (s/? (s/cat :binding-var ::variable-name
-                                                      :sep #{'<-}))
-                                 :fact-type any?
-                                 :destructured-fact (s/? ::destructured)
-                                 :expressions (s/* any?))))
-
-(s/def ::condition (s/or :accumulator ::accumulator
-                         :fact-expr ::fact-expr
-                         :boolean-expr ::boolean-expr
-                         :test ::test))
+                             (s/cat :op #{:and :or :not :exists}
+                                    :conditions (s/+ (s/or :boolean-expr ::boolean-expr
+                                                           :condition ::condition)))))
+
+(s/def ::fact-binding-expr (s/and vector?
+                            (s/cat :fact-binding (s/? (s/cat :binding-var ::variable-name
+                                                             :sep #{'<-}))
+                                   :condition ::condition)))
+
+(s/def ::expression (s/or :accumulator-expr ::accumulator-expr
+                          :boolean-expr ::boolean-expr
+                          :test-expr ::test-expr
+                          :fact-binding-expr ::fact-binding-expr
+                          :condition ::condition))
+
+(s/def ::lhs (s/* ::expression))
+
+(s/def ::query-arg (s/and keyword?
+                          #(str/starts-with? (name %) "?")))
+
 
 (s/def ::defquery-form
   (s/cat :name simple-symbol?
          :docstring (s/? string?)
          :args (s/and vector?
-                      (s/* any?))                           ; TODO not sure what args can be
-         :conditions (s/+ ::condition)))                    ; TODO looks like multiple conditions inside single vector in the doc
+                      (s/* ::query-arg))
+         ;:lhs ::lhs                                               ; TODO. Not sure why but these don't seem equivalent...
```

Comment:
This was a good read: https://medium.com/@roman01la/parsing-dsls-with-predicates-15e0a9a52be

Made me wonder if the `::lhs` would work if it was `(s/spec ::lhs)`

Reply:
Yeah, it's not clear to me why `:lhs ::lhs` wouldn't work either.

Reply:
Tried `(s/spec ::lhs)` and same result (doesn't work) :(